search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,498 @@
+#include <config.h>
+
+#include <gio/gio.h>
+#include <string.h>
+#include <glib/gi18n.h>
+#include <gdk/gdkx.h>
+
+#include <gtk/gtk.h>
+
+#include "yelp-shell-search-provider-generated.h"
+
+#include "libyelp/yelp-document.h"
+#include "libyelp/yelp-uri.h"
+#include "libyelp/yelp-view.h"
+
+#define SEARCH_PROVIDER_INACTIVITY_TIMEOUT 12000 /* milliseconds */
+
+struct _PageData
+{
+    gchar *title;
+    gchar *title_casefold;
+    gchar *desc;
+    gchar *desc_casefold;
+    GIcon *icon;
+};
+
+typedef struct _PageData PageData;
+
+typedef GApplicationClass YelpSearchProviderAppClass;
+typedef struct _YelpSearchProviderApp YelpSearchProviderApp;
+
+struct _YelpSearchProviderApp {
+    GApplication parent;
+    gulong id;
+    guint name_owner_id;
+    GDBusObjectManagerServer *object_manager;
+    YelpShellSearchProvider2 *skeleton;
+
+    YelpDocument *doc;
+    GHashTable *page_data_hash_map;
+};
+
+GType yelp_search_provider_app_get_type (void);
+
+#define YELP_TYPE_SEARCH_PROVIDER_APP yelp_search_provider_app_get_type()
+#define YELP_SEARCH_PROVIDER_APP(obj) \
+      (G_TYPE_CHECK_INSTANCE_CAST((obj), YELP_TYPE_SEARCH_PROVIDER_APP, YelpSearchProviderApp))
+
+G_DEFINE_TYPE (YelpSearchProviderApp, yelp_search_provider_app, G_TYPE_APPLICATION)
+
+static PageData *
+page_data_new_steal (gchar *title,
+                     gchar *desc,
+                     GIcon *icon)
+{
+    PageData *data = g_new0 (PageData, 1);
+    if (title) {
+        data->title = title;
+        data->title_casefold = g_utf8_casefold (title, -1);
+    }
+    if (desc) {
+        data->desc = desc;
+        data->desc_casefold = g_utf8_casefold (desc, -1);
+    }
+    data->icon = icon;
+    return data;
+}
+
+void
+page_data_free (PageData *page_data)
+{
+    g_free (page_data->title);
+    g_free (page_data->title_casefold);
+    g_free (page_data->desc);
+    g_free (page_data->desc_casefold);
+    if (page_data->icon)
+        g_object_unref (page_data->icon);
+    g_free (page_data);
+}
+
+static GVariant *
+get_result_metas (gchar **page_ids,
+                  gpointer user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    GVariantBuilder meta, metas;
+    gchar *gicon_str;
+    gint i;
+
+    g_variant_builder_init (&metas, G_VARIANT_TYPE ("aa{sv}"));
+
+    for (i = 0; page_ids[i] != NULL; i++){
+      PageData *data = g_hash_table_lookup (app->page_data_hash_map, page_ids[i]);
+      g_variant_builder_init (&meta, G_VARIANT_TYPE ("a{sv}"));
+      g_variant_builder_add (&meta, "{sv}",
+                             "id", g_variant_new_string (page_ids[i]));
+      g_variant_builder_add (&meta, "{sv}",
+                             "name", g_variant_new_string (data->title));
+      gicon_str = g_icon_to_string (data->icon);
```

Comment:
I'd do `gicon_str = (data->icon != NULL) ? g_icon_to_string (data->icon) : NULL` to avoid criticals when during preloading we got some invalid icon.